### ISSUE TITLE
TypeCheckException when evaluating a TreeModel with probability output

### ISSUE THREAD

**Issue by yewmun-ngdata** [jpmml/jpmml-evaluator#122]
```
org.jpmml.evaluator.TypeCheckException: Expected org.jpmml.evaluator.HasConfidence value, got org.jpmml.evaluator.tree.TreeModelEvaluator$2 value
	at org.jpmml.evaluator.TypeUtil.cast(TypeUtil.java:502)
	at org.jpmml.evaluator.OutputUtil.getProbability(OutputUtil.java:424)
	at org.jpmml.evaluator.OutputUtil.evaluate(OutputUtil.java:205)
	at org.jpmml.evaluator.tree.TreeModelEvaluator.evaluate(TreeModelEvaluator.java:144)
	at org.jpmml.evaluator.ModelEvaluator.evaluate(ModelEvaluator.java:434)
```
It looks like the TreeModelEvaluator created a org.dmg.pmml.tree.Node which has to be cast to HasProbability in OutputUtil.getProbability(). But Node does not implement the HasProbability interface and hence the TypeCheckException.

**Comment by vruusmann** [comment]
Your PMML document contains a `OutputField@feature="confidence"` instruction, which triggers the "confidence estimation" workflow.

However, as this exception clearly points out, the corresponding `TreeModel` element does not provide sufficient information to do it. Perhaps it's a regression-type decision tree model?

TL;DR: You're dealing with an invalid PMML document. Please talk to the vendor of your PMML exporter software.

**Comment by yewmun-ngdata** [comment]
I copied and paste the wrong stack trace, I set it to feature="confidence" for 1 of my test to see if the same error will happen.

org.jpmml.evaluator.TypeCheckException: Expected org.jpmml.evaluator.HasProbability value, got org.jpmml.evaluator.tree.TreeModelEvaluator$2 value

	at org.jpmml.evaluator.TypeUtil.cast(TypeUtil.java:502)
	at org.jpmml.evaluator.OutputUtil.getProbability(OutputUtil.java:424)
	at org.jpmml.evaluator.OutputUtil.evaluate(OutputUtil.java:205)
	at org.jpmml.evaluator.tree.TreeModelEvaluator.evaluate(TreeModelEvaluator.java:144)
	at org.jpmml.evaluator.ModelEvaluator.evaluate(ModelEvaluator.java:434)


I have no problem with this pmml when I used Zementis. I tried to debug and I can't find anywhere that the Tree Node created can be a HasProbability, so I don't think it's an invalid PMML problem.

**Comment by vruusmann** [comment]
> This is the correct stack trace.

Makes no difference - you're requesting for something, which isn't conceptually possible.

This is specifically a PMML markup-related problem. If it was some PMML feature that isn't supported by the JPMML-Evaluator library, then a different kind of exception would be thrown - for example `UnsupportedElementException` or `UnsupportedAttributeException`.

If you shared a relevant fragment of your PMML document then maybe I could be more specific.

> I have no problem with this pmml when I used Zementis.

Just keep using Zementis then.

**Comment by yewmun-ngdata** [comment]
> Makes no difference - you're requesting for something, which isn't conceptually possible.

Do you mean that PMML do not allow Probability output in a tree model?

My pmml fragment:

```
<TreeModel modelName="DecisionTree" functionName="classification" splitCharacteristic="binarySplit" missingValueStrategy="defaultChild">

    <MiningSchema>
      <MiningField name="incomeCategory" usageType="predicted"/>
      <MiningField name="monthlyIncome" usageType="active"/>
    </MiningSchema>

    <Output>
      <OutputField name="Predicted_incomeCategory" optype="categorical" dataType="string" feature="predictedValue"/>
      <OutputField name="Probability_low" optype="continuous" dataType="double" feature="probability" value="low"/>
      <OutputField name="Probability_average" optype="continuous" dataType="double" feature="probability" value="average"/>
      <OutputField name="Probability_high" optype="continuous" dataType="double" feature="probability" value="high"/>
      <OutputField name="Probability_rich" optype="continuous" dataType="double" feature="probability" value="rich"/>
    </Output>

    <Node id="1" score="average" defaultChild="1.1">
      <True/>
      <Node id="1.1" score="average" defaultChild="1.1.2">
        <SimplePredicate field="monthlyIncome" operator="lessThan" value="4000"/>
        <Node id="1.1.1" score="low">
          <SimplePredicate field="monthlyIncome" operator="lessThan" value="2000"/>
        </Node>
        <Node id="1.1.2" score="average">
          <True/>
        </Node>
      </Node>
      <Node id="1.2" score="high" defaultChild="1.2.2">
        <True/>
        <Node id="1.2.1" score="rich">
          <SimplePredicate field="monthlyIncome" operator="greaterThan" value="10000"/>
        </Node>
        <Node id="1.2.2" score="high">
          <True/>
        </Node>
      </Node>
    </Node>

  </TreeModel>

```


> Just keep using Zementis then.

I don't mean to say which library is better. My statement was meant to be "my PMML used to work". Of course, if you tell me that it really is a problem with my PMML, then I'll just report as it is to my team.

**Comment by vruusmann** [comment]
> Do you mean that PMML do not allow Probability output in a tree model?

There are (at least-) two different kinds of classification-type decision tree models.

The classical one, which supports `OutputField@feature="probability"`, encodes the probability distribution at each leaf `Node` element using a list of `ScoreDistribution` elements:
```xml
<Node score="average">
  <ScoreDistribution value="high" probability="0.1"/>
  <ScoreDistribution value="average" probability="0.7"/>
  <ScoreDistribution value="low" probability="0.2"/>
</Node>
```

The other one, which DOES NOT support `Output@feature="probability"`, simply gives you the value of the most probable category:
```xml
<Node score="average"/>
```

> Of course, if you tell me that it really is a problem with my PMML, then I'll just report as it is to my team.

If you think that your decision tree model is producing a true probability distribution, then you should make it explicit by switching from "voting"-type representation to the "probability distribution"-type representation of leafs.

**Comment by yewmun-ngdata** [comment]
ok. thanks a lot!
